File: Sean的leetcode/Leetcode/ Find Median from Data Stream.cpp
//Two max heap to keep the 2 medium numbers at each top
//add:O(logn)
//find:O(1)
//space:O(n)
class MedianFinder {
public:

    // Adds a number into the data structure.
    void addNum(int num) {
        small.push(num);
        large.push(-small.top());
        small.pop();
        if (large.size() > small.size()) {
            small.push(-large.top());
            large.pop();
        }
        
    }

    // Returns the median of current data stream
    double findMedian() {
        if (small.size() > large.size()) {
            return small.top();
        } else {
            return (small.top() - large.top()) / 2.0;
        }
    }
private:
    priority_queue<int> small;
    priority_queue<int> large;
};

// Your MedianFinder object will be instantiated and called as such:
// MedianFinder mf;
// mf.addNum(1);
// mf.findMedian();

File: Sean的leetcode/python/Find Median from Data Stream.py
###
# Heap
# addNum: O(logn)
# findMedian: O(1)
# Space: O(n)
###
class MedianFinder(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        from Queue import PriorityQueue
        self.st = PriorityQueue()
        self.bt =  PriorityQueue()
        

    def addNum(self, num):
        """
        :type num: int
        :rtype: void
        """
        self.st.put(-num)
        self.bt.put(-self.st.get())
        if self.bt.qsize() > self.st.qsize():
            self.st.put(-self.bt.get())
        

    def findMedian(self):
        """
        :rtype: float
        """
        if self.st.qsize() > self.bt.qsize():
            return -self.st.queue[0]
        else:
            return (-self.st.queue[0] + self.bt.queue[0])/2.0


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()
        
        